### PR TITLE
Add support for vscale vector code generation.

### PR DESCRIPTION
**Pull request by zvookin**
Factored from the [fixed_length_vectors](https://github.com/halide/Halide/tree/fixed_length_vectors) branch to make PRs smaller and easier to review.

This will be used to support the ARM SVE/SVE2 and RISC V Vector architectures. e.g. https://github.com/halide/Halide/pull/6781 .

### CODE REVIEW

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -4751,5 +4776,114 @@ bool CodeGen_LLVM::supports_call_as_float16(const Call *op) const {
...
+    intrin += ".nxv" + std::to_string(scalable->getMinNumElements()) + type_designator + bits_designator;
+    std::vector<llvm::Value *> args;
+    args.push_back(scalable_arg);
+    args.push_back(ConstantInt::get(i64_t, 0));
+
+    return call_intrin(fixed, fixed->getNumElements(), intrin, args, false);
+}
+
+int CodeGen_LLVM::get_vector_num_elements(const llvm::Type *t) {
+#if LLVM_VERSION >= 120
```
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp:4868]: See previous comment

File: src/CodeGen_Internal.cpp [hunk 2]
```diff
@@ -36,18 +37,20 @@ llvm::Type *llvm_type_of(LLVMContext *c, Halide::Type t) {
...
+        llvm::Type *element_type = llvm_type_of(c, t.element_of(), 0);
+        bool scalable = false;
+        int lanes = t.lanes();
+        if (effective_vscale != 0) {
+            int total_bits = t.bits() * t.lanes();
+            scalable = ((total_bits % effective_vscale) == 0);
+            if (scalable) {
+                lanes /= effective_vscale;
+            } else {
+                debug(0) << "Failed to make scalable with bits " << t.bits() << " lanes " << t.lanes()
```
Comment (abadams) [inline src/CodeGen_Internal.cpp:49]: Should this be internal_error?

**Review by abadams — APPROVED** [review]

File: src/CodeGen_Internal.cpp (hunk 2 above)
Comment (zvookin) [reply src/CodeGen_Internal.cpp]: Done. Added comment to document the situation.

**Comment by steven-johnson** [comment]
Is this ready to land?

**Pull request merged**